File: ChatAppCommon/src/github/chatapp/common/message_types/Message.java
/* Copyright (C) 2023 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.chatapp.common.message_types;

import java.util.Arrays;
import java.util.Objects;

/**
 * @author Ilias Koukovinis
 *
 */
public final class Message {
	
	private String username;
	private int clientID;
	
	private int messageID;
	private int chatSessionID;

	private byte[] text;
	private byte[] fileName;
	
	private ContentType contentType;

	public Message() {}

	public Message(String username, int clientID, int messageID, int chatSessionID, byte[] text, byte[] fileName, ContentType contentType) {
		this.username = username;
		this.clientID = clientID;
		this.messageID = messageID;
		this.chatSessionID = chatSessionID;
		this.text = text;
		this.fileName = fileName;
		this.contentType = contentType;
	}

	public void setUsername(String username) {
		this.username = username;
	}

	public void setClientID(int clientID) {
		this.clientID = clientID;
	}

	public void setMessageID(int messageID) {
		this.messageID = messageID;
	}

	public void setChatSessionID(int chatSessionID) {
		this.chatSessionID = chatSessionID;
	}

	public void setText(byte[] text) {
		this.text = text;
	}

	public void setFileName(byte[] fileName) {
		this.fileName = fileName;
	}

	public void setContentType(ContentType contentType) {
		this.contentType = contentType;
	}
	
	public String getUsername() {
		return username;
	}

	public int getClientID() {
		return clientID;
	}

	public int getMessageID() {
		return messageID;
	}

	public int getChatSessionID() {
		return chatSessionID;
	}

	public byte[] getText() {
		return text;
	}

	public byte[] getFileName() {
		return fileName;
	}

	public ContentType getContentType() {
		return contentType;
	}

	@Override
	public int hashCode() {
		return Objects.hashCode(messageID);
	}

	@Override
	public boolean equals(Object obj) {
		
		if (this == obj) {
			return true;
		}
		
		if (obj == null) {
			return false;
		}
		
		if (getClass() != obj.getClass()) {
			return false;
		}
		
		Message other = (Message) obj;
		return chatSessionID == other.chatSessionID && clientID == other.clientID && contentType == other.contentType
				&& Arrays.equals(fileName, other.fileName) && messageID == other.messageID
				&& Arrays.equals(text, other.text) && Objects.equals(username, other.username);
	}

	@Override
	public String toString() {
		return "Message [username=" + username + ", clientID=" + clientID + ", messageID=" + messageID
				+ ", chatSessionID=" + chatSessionID + ", text=" + Arrays.toString(text) + ", fileName="
				+ Arrays.toString(fileName) + ", contentType=" + contentType + "]";
	}

}

File: ErmisClient/Mobile/ermis_client/lib/client/common/message.dart
/* Copyright (C) 2024 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import 'dart:typed_data';

import 'common.dart';

class Message {
  late String username;
  late int clientID;
  late int messageID;
  late int chatSessionID;
  late int chatSessionIndex;
  Uint8List? text;
  Uint8List? fileName;
  Uint8List? imageBytes;
  late int timeWritten;
  late ContentType contentType;
  bool isSent = false;

  Message(
      {required this.username,
      required this.clientID,
      required this.messageID,
      required this.chatSessionID,
      required this.chatSessionIndex,
      this.text,
      this.fileName,
      required this.timeWritten,
      required this.contentType,
      required this.isSent});

  Message.empty();

  void setUsername(String username) => this.username = username;
  void setIsSent(bool isSent) => this.isSent = isSent;
  void setClientID(int clientID) => this.clientID = clientID;
  void setMessageID(int messageID) => this.messageID = messageID;
  void setChatSessionID(int chatSessionID) => this.chatSessionID = chatSessionID;
  void setChatSessionIndex(int chatSessionIndex) => this.chatSessionIndex = chatSessionIndex;
  void setText(Uint8List? text) => this.text = text;
  void setFileName(Uint8List? fileName) => this.fileName = fileName;
  void setTimeWritten(int timeWritten) => this.timeWritten = timeWritten;
  void setContentType(ContentType contentType) => this.contentType = contentType;

  String get getUsername => username;
  int get getClientID => clientID;
  int get getMessageID => messageID;
  int get getChatSessionID => chatSessionID;
  int get getChatSessionIndex => chatSessionIndex;
  Uint8List? get getText => text;
  Uint8List? get getFileName => fileName;
  int get getTimeWritten => timeWritten;
  ContentType get getContentType => contentType;
  bool get getIsSent => isSent;

  @override
  int get hashCode => messageID.hashCode;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Message) return false;
    return chatSessionID == other.chatSessionID &&
        chatSessionIndex == other.chatSessionIndex &&
        clientID == other.clientID &&
        contentType == other.contentType &&
        messageID == other.messageID &&
        text == other.text &&
        fileName == other.fileName &&
        timeWritten == other.timeWritten &&
        username == other.username;
  }

  @override
  String toString() {
    return 'Message{username: $username, clientID: $clientID, messageID: $messageID, chatSessionID: $chatSessionID, '
        'chatSessionIndex $chatSessionIndex, text: $text, fileName: $fileName, timeWritten: $timeWritten, contentType: $contentType}';
  }
}
